File: mobile/src/main/java/treehou/se/habit/util/DatabaseServerLoaderFactory.java
package treehou.se.habit.util;

import android.content.Context;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.ObservableTransformer;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import io.realm.Realm;
import se.treehou.ng.ohcommunicator.connector.models.OHServer;
import se.treehou.ng.ohcommunicator.connector.models.OHSitemap;
import se.treehou.ng.ohcommunicator.services.IServerHandler;
import treehou.se.habit.core.db.OHRealm;
import treehou.se.habit.core.db.model.ServerDB;
import treehou.se.habit.core.db.model.SitemapDB;
import treehou.se.habit.module.ServerLoaderFactory;

public class DatabaseServerLoaderFactory implements ServerLoaderFactory {

    private static final String TAG = DatabaseServerLoaderFactory.class.getSimpleName();

    private OHRealm realm;
    private ConnectionFactory connectionFactory;

    @Inject
    public DatabaseServerLoaderFactory(OHRealm realm, ConnectionFactory connectionFactory) {
        this.realm = realm;
        this.connectionFactory = connectionFactory;
    }

    @Override
    public OHServer loadServer(Realm realm, long serverId) {
        return ServerDB.Companion.load(realm, serverId).toGeneric();
    }

    @Override
    public ObservableTransformer<Realm, OHServer> loadServersRx() {
        return RxUtil.loadServers();
    }

    @Override
    public ObservableTransformer<Realm, List<OHServer>> loadAllServersRx() {
        return observable -> observable.flatMap(realmLocal ->
                realmLocal.where(ServerDB.class).isNotEmpty("localurl").or().isNotEmpty("remoteurl").greaterThan("id", 0).findAllAsync().asFlowable().toObservable())
                .map(serverDBS -> {
                    List<OHServer> serverList = new ArrayList<>();
                    for(ServerDB serverDB : serverDBS){
                        serverList.add(serverDB.toGeneric());
                    }
                    return serverList;
                })
                .distinct();
    }

    @Override
    public ObservableTransformer<List<OHServer>, List<ServerSitemapsResponse>> serversToSitemap(Context context) {
        return observable -> observable
                .switchMap(servers -> {
                    List<Observable<ServerSitemapsResponse>> sitemapResponseRx = new ArrayList<>();
                    for (OHServer server : servers) {
                        Observable<ServerSitemapsResponse> serverSitemapsResponseRx = Observable.just(server)
                                .compose(serverToSitemap(context))
                                .subscribeOn(Schedulers.io())
                                .startWith(Companion.getEMPTY_RESPONSE());

                        sitemapResponseRx.add(serverSitemapsResponseRx);
                    }

                    return Observable.<ServerSitemapsResponse, List<ServerSitemapsResponse>>combineLatest(sitemapResponseRx, responsesObject -> {
                        List<ServerSitemapsResponse> responses = new ArrayList<>();
                        for (Object responseObject : responsesObject) {
                            ServerSitemapsResponse response = (ServerSitemapsResponse) responseObject;
                            responses.add(response);
                        }

                        return responses;
                    });
                });
    }

    /**
     * Fetches sitemaps from server.
     * @param context the used to fetch sitemaps.
     * @return
     */
    public ObservableTransformer<OHServer, ServerSitemapsResponse> serverToSitemap(Context context) {
        return observable -> observable
                .flatMap(server -> {
            IServerHandler serverHandler = connectionFactory.createServerHandler(server, context);
            return serverHandler.requestSitemapRx()
                    .map(SitemapResponse::new)
                    .subscribeOn(Schedulers.io())
                    .doOnError(e -> Log.e(TAG, "Failed to load sitemap", e))
                    .onErrorReturn(throwable -> new SitemapResponse(new ArrayList<>(), throwable));
        }, (server, sitemapResponse) -> {
            for (OHSitemap sitemap : sitemapResponse.sitemaps) {
                sitemap.setServer(server);
            }
            return new ServerSitemapsResponse(server, sitemapResponse.sitemaps, sitemapResponse.error);
        })
        .doOnNext(RxUtil.saveSitemap());
    }

    public ObservableTransformer<ServerSitemapsResponse, ServerSitemapsResponse> filterDisplaySitemaps() {
        return observable -> observable.map((Function<ServerSitemapsResponse, ServerSitemapsResponse>) this::filterDisplaySitemaps);
    }

    @Override
    public ObservableTransformer<List<ServerSitemapsResponse>, List<ServerSitemapsResponse>> filterDisplaySitemapsList() {
        return observable -> observable.map((Function<List<ServerSitemapsResponse>, List<ServerSitemapsResponse>>) serverSitemapsResponses -> {
            List<ServerSitemapsResponse> responses = new ArrayList<>();
            for (ServerSitemapsResponse sitemapsResponse : serverSitemapsResponses){
                responses.add(filterDisplaySitemaps(sitemapsResponse));
            }
            return responses;
        });
    }

    private ServerSitemapsResponse filterDisplaySitemaps(ServerSitemapsResponse response){
        List<OHSitemap> sitemaps = new ArrayList<>();
        for(OHSitemap sitemap : response.getSitemaps()){
            SitemapDB sitemapDB = realm.realm().where(SitemapDB.class)
                    .equalTo("name", sitemap.getName())
                    .equalTo("server.name", response.getServer().getName())
                    .findFirst();

            if(sitemapDB == null || sitemapDB.getSettingsDB() == null
                    || sitemapDB.getSettingsDB().getDisplay()){

                sitemaps.add(sitemap);
            }
        }
        return new ServerSitemapsResponse(response.getServer(), sitemaps, response.getError());
    }

    private static class SitemapResponse{
        public Throwable error;
        public List<OHSitemap> sitemaps;

        public SitemapResponse(List<OHSitemap> sitemaps) {
            this.sitemaps = sitemaps;
        }

        public SitemapResponse(List<OHSitemap> sitemaps, Throwable error) {
            this.error = error;
            this.sitemaps = sitemaps;
        }
    }
}

File: mobile/src/main/java/treehou/se/habit/util/DatabaseServerLoaderFactory.kt
package treehou.se.habit.util

import android.content.Context
import android.util.Log
import io.reactivex.Observable
import io.reactivex.ObservableTransformer
import io.reactivex.functions.Function
import io.reactivex.schedulers.Schedulers
import io.realm.Realm
import se.treehou.ng.ohcommunicator.connector.models.OHServer
import se.treehou.ng.ohcommunicator.connector.models.OHSitemap
import treehou.se.habit.core.db.OHRealm
import treehou.se.habit.core.db.model.ServerDB
import treehou.se.habit.core.db.model.SitemapDB
import treehou.se.habit.dagger.ServerLoaderFactory
import java.util.*
import javax.inject.Inject

class DatabaseServerLoaderFactory @Inject
constructor(private val realm: OHRealm, private val connectionFactory: ConnectionFactory) : ServerLoaderFactory {

    override fun loadServer(realm: Realm, id: Long): OHServer {
        return ServerDB.load(realm, id)!!.toGeneric()
    }

    override fun loadServersRx(): ObservableTransformer<Realm, OHServer> {
        return RxUtil.loadServers()
    }

    override fun loadAllServersRx(): ObservableTransformer<Realm, List<OHServer>> {

        return ObservableTransformer { observable ->
            observable.flatMap({ realmLocal ->
                realmLocal.where(ServerDB::class.java).isNotEmpty("localurl").or().isNotEmpty("remoteurl").greaterThan("id", 0).findAllAsync()
                        .asFlowable().toObservable()
            })
                    .map({ serverDBS ->
                        serverDBS.map { it.toGeneric() }
                    })
                    .distinct()
        }
    }

    override fun serversToSitemap(context: Context?): ObservableTransformer<List<OHServer>, List<ServerLoaderFactory.ServerSitemapsResponse>> {
        return ObservableTransformer { observable ->
            observable
                    .switchMap({ servers ->
                        val sitemapResponseRx = ArrayList<Observable<ServerLoaderFactory.ServerSitemapsResponse>>()
                        for (server in servers) {
                            val serverSitemapsResponseRx = Observable.just<OHServer>(server)
                                    .compose<ServerLoaderFactory.ServerSitemapsResponse>(serverToSitemap(context!!))
                                    .subscribeOn(Schedulers.io())
                                    .startWith(ServerLoaderFactory.Companion.EMPTY_RESPONSE)

                            sitemapResponseRx.add(serverSitemapsResponseRx)
                        }

                        Observable.combineLatest<ServerLoaderFactory.ServerSitemapsResponse, List<ServerLoaderFactory.ServerSitemapsResponse>>(sitemapResponseRx) { responsesObject ->
                            val responses = ArrayList<ServerLoaderFactory.ServerSitemapsResponse>()
                            for (responseObject in responsesObject) {
                                val response = responseObject as ServerLoaderFactory.ServerSitemapsResponse
                                responses.add(response)
                            }

                            responses
                        }
                    })
        }
    }

    /**
     * Fetches sitemaps from server.
     * @param context the used to fetch sitemaps.
     * @return
     */
    override fun serverToSitemap(context: Context): ObservableTransformer<OHServer, ServerLoaderFactory.ServerSitemapsResponse> {
        return ObservableTransformer { observable ->
            observable
                    .flatMap({ server ->
                        val serverHandler = connectionFactory.createServerHandler(server, context)
                        serverHandler.requestSitemapRx()
                                .map { SitemapResponse(it) }
                                .subscribeOn(Schedulers.io())
                                .doOnError { e -> Log.e(TAG, "Failed to load sitemap", e) }
                                .onErrorReturn { throwable -> SitemapResponse(ArrayList(), throwable) }
                    }, { server, sitemapResponse ->
                        for (sitemap in sitemapResponse.sitemaps) {
                            sitemap.setServer(server)
                        }
                        ServerLoaderFactory.ServerSitemapsResponse(server, sitemapResponse.sitemaps, sitemapResponse.error)
                    })
                    .doOnNext(RxUtil.saveSitemap())
        }
    }

    override fun filterDisplaySitemaps(): ObservableTransformer<ServerLoaderFactory.ServerSitemapsResponse, ServerLoaderFactory.ServerSitemapsResponse> {
        return ObservableTransformer { observable -> observable.map(Function<ServerLoaderFactory.ServerSitemapsResponse, ServerLoaderFactory.ServerSitemapsResponse> { this.filterDisplaySitemaps(it) }) }
    }

    override fun filterDisplaySitemapsList(): ObservableTransformer<List<ServerLoaderFactory.ServerSitemapsResponse>, List<ServerLoaderFactory.ServerSitemapsResponse>> {
        return ObservableTransformer { observable ->
            observable.map({ serverSitemapsResponses: List<ServerLoaderFactory.ServerSitemapsResponse> ->
                val responses = ArrayList<ServerLoaderFactory.ServerSitemapsResponse>()
                for (sitemapsResponse in serverSitemapsResponses) {
                    responses.add(filterDisplaySitemaps(sitemapsResponse))
                }
                responses
            })
        }
    }

    private fun filterDisplaySitemaps(response: ServerLoaderFactory.ServerSitemapsResponse): ServerLoaderFactory.ServerSitemapsResponse {
        val sitemaps = ArrayList<OHSitemap>()
        for (sitemap in response.sitemaps!!) {
            val sitemapDB = realm.realm().where(SitemapDB::class.java)
                    .equalTo("name", sitemap.name)
                    .equalTo("server.name", response.server!!.name)
                    .findFirst()

            if (sitemapDB == null || sitemapDB.settingsDB == null
                    || sitemapDB.settingsDB!!.display) {

                sitemaps.add(sitemap)
            }
        }
        return ServerLoaderFactory.ServerSitemapsResponse(response.server, sitemaps, response.error)
    }

    private class SitemapResponse {
        var error: Throwable? = null
        var sitemaps: List<OHSitemap>

        constructor(sitemaps: List<OHSitemap>) {
            this.sitemaps = sitemaps
        }

        constructor(sitemaps: List<OHSitemap>, error: Throwable) {
            this.error = error
            this.sitemaps = sitemaps
        }
    }

    companion object {

        private val TAG = DatabaseServerLoaderFactory::class.java.simpleName
    }
}
